### ISSUE TITLE
Show admin warning if number of outstanding jobs > 200

### ISSUE THREAD

**Issue by user23498723452** [nextcloud/server#18582]
I use NC 16.0.6.1 on a QNAP NAS (proprietary linux OS).  I moved the log files so I could see them remotely, but it seems that the internal log rotation method is somehow incompatible. 

This could be an issue with the OS, or it could be an issue with the log rotation code.  There is nothing logged in the nextcloud.log that is related (and no errors either)

As best I can tell,  my webserver (and thus NC) runs as admin (which is root on QNAP)

**config.php:**
```
  'logtimezone' => 'America/New_York',
  'log_type' => 'file',
  'logfile' => '/share/Download/nextcloud/nextcloud.log',
  'loglevel' => 0,
  'log_rotate_size' => 10204857,

```

**log dir:**
```
[/share/Download/nextcloud] # ls -altr
total 259516
-rw-r----- 1 admin administrators  61695837 2019-07-21 10:24 audit.log_7-21-2019
-rw-r----- 1 admin administrators    425969 2019-09-17 23:30 nextcloud.log.9-18-2019
-rw-r----- 1 admin administrators   7365846 2019-09-17 23:30 audit.log.9-18-2019
drwxrwxrwx 9 admin administrators     36864 2019-12-24 17:01 ../
-rw-r----- 1 admin administrators  10139861 2019-12-27 01:19 audit.log
drwxr-xr-x 2 admin everyone            4096 2019-12-27 10:32 ./
-rw-r----- 1 admin administrators 185993715 2019-12-27 10:44 nextcloud.log
```

**nextcloud.log:**
```

{"reqId":"NRyqDmz2X9pf2kIcbAJJ","level":0,"time":"2019-12-27T11:02:38-05:00","remoteAddr":"192.168.1.246","user":"--","app":"cron","method":"GET","url":"\/cron.php","message":"Run OC\\Command\\CommandJob job with ID 1936","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko\/20100101 Firefox\/71.0","version":"16.0.6.1"}
{"reqId":"NRyqDmz2X9pf2kIcbAJJ","level":0,"time":"2019-12-27T11:02:38-05:00","remoteAddr":"192.168.1.246","user":"--","app":"cron","method":"GET","url":"\/cron.php","message":"Finished OC\\Command\\CommandJob job with ID 1936 in 0 seconds","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko\/20100101 Firefox\/71.0","version":"16.0.6.1"}
{"reqId":"r3cOXNUhOy5A10ZkRyHv","level":0,"time":"2019-12-27T11:07:27-05:00","remoteAddr":"192.168.1.246","user":"--","app":"cron","method":"GET","url":"\/cron.php","message":"Run OC\\Command\\CommandJob job with ID 1938","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko\/20100101 Firefox\/71.0","version":"16.0.6.1"}
{"reqId":"r3cOXNUhOy5A10ZkRyHv","level":0,"time":"2019-12-27T11:07:27-05:00","remoteAddr":"192.168.1.246","user":"--","app":"cron","method":"GET","url":"\/cron.php","message":"Finished OC\\Command\\CommandJob job with ID 1938 in 0 seconds","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:71.0) Gecko\/20100101 Firefox\/71.0","version":"16.0.6.1"}

```

**Comment by kesselb** [comment]
Nextcloud keep track of executed jobs in `oc_jobs` table. What does it show for `OC\Log\Rotate`?

**Comment by user23498723452** [comment]
I'm not great with the db stuff. A row appears for OC\Log\Rotate,  and the epoch times are up to date.
let me know if you need more info. 

![Selection_230](https://user-images.githubusercontent.com/17534291/71525539-13c86500-28a0-11ea-99e5-c7f20ca11378.png)

**Comment by user23498723452** [comment]
To be clear,  the log files are not near the web app dir.  (But they are owned/accessed by the same user).

**Comment by user23498723452** [comment]
I do have over 5000 rows in that table.  Most appear to be related to Expiration of file versions. 

`"O:33:\"OCA\\Files_Versions\\Command\\Expire\":2:{s:43:\"\u0000OCA\\Files_Versions\\Command\\Expire\u0000fileName\";s:57:\"\/InstantUpload\/Screenshots\/Screenshot_20190409-130619.png\";s:39:\"\u0000OCA\\Files_Versions\\Command\\Expire\u0000user\";s:5:\"user\";}"`

**Comment by kesselb** [comment]
https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/Log/Rotate.php#L40

Code in question is using `logfile` if present or `datadirectory/nextcloud.log` otherwise. So this should be fine. Timestamp: 1546724017 => Date: Sat, 05 Jan 2019 22:33:37 +0100. That's a while. Are the other jobs having a newer timestamp? If not something with the cron.php do not work. Usually there is a warning with the admin interface if the last cron execution is overdue.

**Comment by user23498723452** [comment]
My bad.  I'm apparently using AJAX for the cron method.  Does this prevent the tasks as I'm seeing?

**Comment by kesselb** [comment]
> I do have over 5000 rows in that table. Most appear to be related to Expiration of file versions.

Oh. `cron.php` is triggered if you're using nextcloud. To not execute the webservers `max_execution_time` only one job is executed with this method. So your probably have a huge backlog of jobs :fearful: Try to run php cron.php via ssh.

**Comment by kesselb** [comment]
I was not aware that files_expiration could create so many background jobs. We should probably add a setupcheck that warns a user about such a situation. Like "oh. you're having 5000 jobs in the queue. please use system cron for such setups". Or remove these ajax method because it will never scale on bigger setups anyway.

cc @rullzer @nickvergessen

**Comment by user23498723452** [comment]
It seems that the default AJAX method must only work for interactive web sessions?  Because I have clients that regularly talk to the server.  Seems like they would cull the request list. (but I could be wrong)

I've just tried to enable the cron method.  It runs every 5 min.  I will let you know what happens.

**Comment by user23498723452** [comment]
It appears that the logfile rotated now.  The root cause was the cron tasks were not all being executed.
I would argue that some of those tasks are more important than others.  Perhaps some prioritization is required to prevent issues like this?  As opposed to telling all users that they simply can't do this (log rotation) unless they are a linux expert?
It would seem to be more user friendly to simply adapt the code to account for "lower priority" tasks like deleting old versions of files. And ensure that "core system" tasks are executed more consistently (instead of assuming they will eventually run)

**Comment by nickvergessen** [comment]
Well ajax works out of the box and is suitable for one/two user instances. Anything else should use regular/web cron. I think the setting in the UI shows that already

**Comment by kesselb** [comment]
> Logrotation problem - ajax cron method preventing it from occurring

That's missleading ;)

> It seems that the default AJAX method must only work for interactive web sessions?

That' true and is exactly the issue here with the huge backlog as result.

> Well ajax works out of the box and is suitable for one/two user instances.

Yes and no. I'm using the desktop and android client with my personal instance. I'm visiting the web ui probably once a month. So ajax method would not work on my instance. But I don't know how to do it better. 

> simply adapt the code to account for "lower priority" tasks like deleting old versions of files.

I disagree. To much work for edge case. Also the priority depends on your use case.

**Comment by user23498723452** [comment]
> Well ajax works out of the box and is suitable for one/two user instances.

I disagree.  I am a single user.  It is not working for me.  The issue is around the amount of data and changes to that data.  

> I disagree. To much work for edge case. Also the priority depends on your use case.

Fair point.

That said, I have gotten past the issue.  But I think the value of making the AJAX the default should be seriously reconsidered, and the right thing to do is surface some UI element to users that are having issues like mine.  This is very detectable and very preventable.  

I do think the AJAX method could be improved with some reasonable assumptions about task handling.  Like why can't you just run core tasks that are present by default first.  Then run other tasks after that. Why can't you check for the oldest task and run that one. (maybe you are?) 
Finally, a single task is executed?  You can probably identify the timeout scenario you mentioned when running lots of tasks, via some simple flag ("running_tasks=true"), then take remedial action if it is found at the start of the next run.  You could also do some math during task execution to determine if the count is going up or down, and use that math to calculate how many to try the NEXT time it runs.

I'm not saying this is trivial,  but it isn't _that_ hard.  And I think users deserve this to prevent any number of issues - and who knows - maybe fix a lot of ongoing issues!

**Comment by nickvergessen** [comment]
Well I totally disagree there.

Yes we should make users aware if the backlog of jobs is growing too big. But for bigger instances there is just no way around normal cron. If you only sync via the clients and never open the UI the list of background jobs also only ever gets bigger. Sure we could fire another request in ajax cron if there are more tasks to run, but it only works if someone visits the web and keeps the tab open.

**Comment by user23498723452** [comment]
UPDATE:  I may not have solved my issue on QNAP.  The OS decided to comment out my valid crontab entry.  As I stated, they do some weird stuff.  I have manually re-enabled it,  but cannot count on this going forward is my guess.

Again,  I implore the product team to consider improving native functionality to eliminate the need to for eternal tooling.

**Comment by kesselb** [comment]
> I implore the product team to consider improving native functionality to eliminate the need to for external tooling.

1) Do you have any suggestion how to improve this?
2) Why do you think someone should spend time on this?

I'm not against it in general. But I don't see any quick wins. We could probably measure the average execution time of each job and do some guessing like (max_execution_time - time passed) / average_job_execution_time use the lower value and hope the webserver will not kill the job. 

Spending time in improving the ajax cron method seems like reinventing the wheel (check this great article https://sourcemaking.com/antipatterns/reinvent-the-wheel). Wikipedia says that `cron` has been released by AT&T Bell Laboratories in May 1975. That's 44 years ago! Cron will do one thing. Executing a given command to a fixed time. The execution time is unlimited (in most environments). That's amazing because all the expensive and time consuming jobs can be run. No webserver will terminate request due script running times. 

> I may not have solved my issue on QNAP. The OS decided to comment out my valid crontab entry. 

That sucks! Nextcloud has many issues but I'm pretty certain that we are not commenting any crontab entries ;) 

tl;dr Sorry I don't see that coming any time soon but looking forward to your suggestions.

**Comment by user23498723452** [comment]
I discovered the source of the commented cron job.  I had to disable a security agent on the NAS.  This is concerning of course, and I'd rather not do it, but there are no alternatives on the platform.

Cron jobs are a common malware reinfection technique on linux systems.  That is why they disable non-standard jobs.

I think my point about apps being self-sufficient is my general feedback. Relying on external (unpredictable) system components can only weaken your products reliability.   Making them an option is much wiser.  But flat out requiring them is bad design IMHO. Because who knows how long it takes someone to figure out that they NEED to implement it, as well as if & how to implement it.

**Comment by nickvergessen** [comment]
The admin interface shows you a red warning when the cron wasn't executed for a long time. Also it has a link to the docs how to switch to system cron.

I think we should extend the check to the number of queued cron tasks (e.g. over 100 or 200) to warn admins when the ajax cron is not sufficient anymore (could happen with versions/trashbin)

But we are not going to write a fake replacement for real system cron because its simply impossible
